docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md

```diff
@@ -0,0 +1,505 @@
+
+## Build data driven tests
+
+Your analyzer and code fix work on a simple case of a single declaration that can be made const. There are numerous possible declaration statements where this implementation makes mistakes. You'll address these cases by working with the unit test library written by the template. It's much faster than repeatedly opening a second copy of Visual Studio.
+
+Open the **MakeConstUnitTests.cs** file in the unit test project. The template created two tests that follow the two common patterns for an analyzer and code fix unit test. `Test1` shows the pattern for a test that ensures the analyzer doesn't report a diagnostic when it shouldn't. `Test2` shows the pattern for reporting a diagnostic and running the code fix.
+
+The code for almost every test for your analyzer follows one of these two patterns. For the first step, you can rework these tests as data driven tests. Then, it will be easy to create new tests by adding new string constants to represent different test inputs.
+
+For efficiency, the first step is to refactor the two tests into data driven tests. Then, you only need to define a couple string constants for each new test. While your refactoring, rename both methods to better names. Replace `Test1` with this test that ensures no diagnostic is raised:
+
+```csharp
+[DataTestMethod]
+[DataRow("")]
+public void WhenTestCodeIsValidNoDiagnosticIsTriggered(string testCode)
+{
+    VerifyCSharpDiagnostic(testCode);
+}
+```
+
+You can create a new data row for this test by defining any code fragment that should not cause your diagnostic to trigger a warning. This overload of `VerifyCSharpDiagnostic` passes when there are no diagnostics triggered for the source code fragment.  
+
+Next, replace `Test2` with this test that ensures a diagnostic is raised and a code fix applied for the source code fragment:
+
+```csharp
+[DataTestMethod]
+[DataRow(LocalIntCouldBeConstant, LocalIntCouldBeConstantFixed, 10, 13)]
+public void WhenDiagosticIsRaisedFixUpdatesCode(
+    string test,
+    string fixTest,
+    int line,
+    int column)
+{
+    var expected = new DiagnosticResult
+    {
+        Id = MakeConstAnalyzer.DiagnosticId,
+        Message = new LocalizableResourceString(nameof(MakeConst.Resources.AnalyzerMessageFormat), MakeConst.Resources.ResourceManager, typeof(MakeConst.Resources)).ToString(),
+        Severity = DiagnosticSeverity.Warning,
+        Locations =
+            new[] {
+                    new DiagnosticResultLocation("Test0.cs", line, column)
+                }
+    };
+
+    VerifyCSharpDiagnostic(test, expected);
+
+    VerifyCSharpFix(test, fixTest);
+}
+```
+
+The preceding code also made a couple changes to the code that builds the expected diagnostic result. It uses the public constants registered in the `MakeConst` analyzer. In addition, it uses two string constants for the input and fixed source:
```

Comment:
You don't specify that the two strings should be pasted at the top of the test cs file.